src/backend/commands/analyze.c

```diff
@@ -2625,66 +2549,93 @@ acquire_sample_rows_dispatcher(Relation onerel, bool inh, int elevel,
+	dvalues = (Datum *) palloc0(relDesc->natts * sizeof(Datum));
+	dnulls = (bool *) palloc(relDesc->natts * sizeof(bool));
 	sampleTuples = 0;
 	*totalrows = 0;
 	*totaldeadrows = 0;
-	for (int resultno = 0; resultno < cdb_pgresults.numResults; resultno++)
+
+	slot = MakeSingleTupleTableSlot(queryDesc->tupDesc);
+	while (tuplestore_gettupleslot(portal->holdStore, true, false, slot))
 	{
-		struct pg_result *pgresult = cdb_pgresults.pg_results[resultno];
-		bool		got_summary = false;
+		TupleDesc	typeinfo = slot->tts_tupleDescriptor;
+		int			natts = typeinfo->natts;
+		Datum		attr;
+		bool		isnull;
 		double		this_totalrows = 0;
 		double		this_totaldeadrows = 0;
 
-		if (PQresultStatus(pgresult) != PGRES_TUPLES_OK)
+		/* There should be only one attribute with OID RECORDOID. */
+		if (1 != natts)
 		{
-			cdbdisp_clearCdbPgResults(&cdb_pgresults);
-			ereport(ERROR,
-					(errmsg("unexpected result from segment: %d",
-							PQresultStatus(pgresult))));
+			elog(ERROR,
+				 "wrong number of attributes %d when 1 expected",
+				 natts);
 		}
+		if (RECORDOID != typeinfo->attrs[0]->atttypid)
+		{
+			elog(ERROR,
+				 "wrong attribute OID %d, RECORDOID %d is expected",
+				 typeinfo->attrs[0]->atttypid, RECORDOID);
 
-		if (GpPolicyIsReplicated(onerel->rd_cdbpolicy))
+		}
+
+		/* Make sure the tuple is fully deconstructed */
+		slot_getallattrs(slot);
+
+		/* There should be only one attribute with OID RECORDOID */
+		attr = slot_getattr(slot, 1, &isnull);
+		if (isnull)
 		{
-			/*
-			 * A replicated table has the same data in all segments. Arbitrarily,
-			 * use the sample from the first segment, and discard the rest.
-			 * (This is rather inefficient, of course. It would be better to
-			 * dispatch to only one segment, but there is no easy API for that
-			 * in the dispatcher.)
-			 */
-			if (resultno > 0)
-				continue;
+			elog(ERROR,
+				 "null value for attribute in tuple");
 		}
 
-		for (int rowno = 0; rowno < PQntuples(pgresult); rowno++)
+		/* Get record from attribute and parse it */
 		{
-			/*
-			 * We cannot use record_in function to get row record here.
-			 * Since the result row may contain just the totalrows info where the data columns
-			 * are NULLs. Consider domain: 'create domain dnotnull varchar(15) NOT NULL;'
-			 * NULLs are not allowed in data columns.
-			 */
-			char * rowStr = PQgetvalue(pgresult, rowno, 0);
+			HeapTupleHeader rec = (HeapTupleHeader) PG_DETOAST_DATUM(attr);
+			Oid			tupType;
+			int32		tupTypmod;
+			TupleDesc	tupdesc;
+			HeapTupleData tuple;
+
+			/* Extract type info from the tuple itself */
+			tupType = HeapTupleHeaderGetTypeId(rec);
+			tupTypmod = HeapTupleHeaderGetTypMod(rec);
+			tupdesc = lookup_rowtype_tupdesc(tupType, tupTypmod);
 
-			if (rowStr == NULL)
-				elog(ERROR, "got NULL pointer from return value of gp_acquire_sample_rows");
+			/* Build a temporary HeapTuple control structure */
+			tuple.t_len = HeapTupleHeaderGetDatumLength(rec);
```

Comment:
@seryozhasmirnov I've tried to use memory tuples here but failed to deform correct results. Do you know where I have made an error it this patch?
```
diff --git a/src/backend/commands/analyze.c b/src/backend/commands/analyze.c
index 71c7c006ba2..ac7257203ef 100644
--- a/src/backend/commands/analyze.c
+++ b/src/backend/commands/analyze.c
@@ -2599,20 +2599,18 @@ acquire_sample_rows_dispatcher(Relation onerel, bool inh, int elevel,
                        Oid                     tupType;
                        int32           tupTypmod;
                        TupleDesc       tupdesc;
-                       HeapTupleData tuple;
+                       MemTuple        mtup;
+                       MemTupleBinding *mt_bind;
 
                        /* Extract type info from the tuple itself */
                        tupType = HeapTupleHeaderGetTypeId(rec);
                        tupTypmod = HeapTupleHeaderGetTypMod(rec);
                        tupdesc = lookup_rowtype_tupdesc(tupType, tupTypmod);
 
-                       /* Build a temporary HeapTuple control structure */
-                       tuple.t_len = HeapTupleHeaderGetDatumLength(rec);
-                       ItemPointerSetInvalid(&(tuple.t_self));
-                       tuple.t_data = rec;
-
                        /* Break down the tuple into fields */
-                       heap_deform_tuple(&tuple, tupdesc, funcRetValues, funcRetNulls);
+                       mtup = ExecFetchSlotMemTuple(slot);
+                       mt_bind = create_memtuple_binding(tupdesc);
+                       memtuple_deform(mtup, mt_bind, funcRetValues, funcRetNulls);
 
                        if (!funcRetNulls[0])
                        {
```